spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkScan.java

```diff
@@ -101,16 +99,14 @@ protected List<Expression> filterExpressions() {
     return filterExpressions;
   }
 
-  protected abstract List<CombinedScanTask> tasks();
-
   @Override
   public Batch toBatch() {
-    return new SparkBatch(sparkContext, table, readConf, tasks(), expectedSchema);
+    return this;
```

Comment:
Nice !, this is in sync with built in v2 file Sources: https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala#L197

Reply:
My suspicion is that there is some check that the scan implements Batch, or some other equality check someplace, that is required for filter pushdown.